reactivesocket-client/src/main/java/io/reactivesocket/client/LoadBalancer.java

```diff
@@ -121,18 +179,59 @@ public LoadBalancer(Publisher<List<ReactiveSocketFactory<SocketAddress>>> factor
     }
 
     private synchronized void addSockets(int numberOfNewSocket) {
-        activeFactories.entrySet()
-            .stream()
-            // available factories that don't map to an already established socket
-            .filter(e -> !activeSockets.containsKey(e.getKey()))
-            .map(e -> e.getValue())
-            .filter(factory -> factory.availability() > 0.0)
-            .sorted((a, b) -> -Double.compare(a.availability(), b.availability()))
-            .limit(numberOfNewSocket)
-            .forEach(factory -> {
-                pendingSockets += 1;
-                factory.apply().subscribe(new SocketAdder(factory.remote()));
-            });
+        int n = numberOfNewSocket;
+        if (n > activeFactories.size()) {
+            n = activeFactories.size();
+            logger.info("addSockets(" + numberOfNewSocket
+                + ") restricted by the number of factories, i.e. addSockets(" + n + ")");
+        }
+
+        Random rng = ThreadLocalRandom.current();
+        while (n > 0) {
+            int size = activeFactories.size();
+            if (size == 1) {
+                ReactiveSocketFactory<T> factory = activeFactories.get(0);
+                if (factory.availability() > 0.0) {
+                    activeFactories.remove(0);
+                    pendingSockets++;
+                    factory.apply().subscribe(new SocketAdder(factory));
+                }
+                break;
+            }
+            ReactiveSocketFactory<T> factory0 = null;
+            ReactiveSocketFactory<T> factory1 = null;
+            int i0 = 0;
+            int i1 = 0;
+            for (int i = 0; i < EFFORT; i++) {
+                i0 = rng.nextInt(size);
+                i1 = rng.nextInt(size - 1);
+                if (i1 >= i0) {
+                    i1++;
+                }
+                factory0 = activeFactories.get(i0);
+                factory1 = activeFactories.get(i1);
+                if (factory0.availability() > 0.0 && factory1.availability() > 0.0)
+                    break;
+            }
+
+            if (factory0.availability() < factory1.availability()) {
+                n--;
+                pendingSockets++;
+                if (i1 < size - 1) {
```

Comment:
I think it will be good to put a comment here as to why we are doing this as opposed to doing `.remove(i1)`.